src/relay/transforms/annotate_texture_storage.cc

```diff
@@ -365,43 +428,85 @@ class RewriteVDStorageScopes : public transform::DeviceAwareExprMutator {
       : transform::DeviceAwareExprMutator(Optional<IRModule>()), storage_scope_(storage_scope) {}
 
   Function Rewrite(const Expr& expr) { return Downcast<Function>(Mutate(expr)); }
 
   Expr VisitExpr_(const VarNode* vn) final {
     if (storage_scope_.find(GetRef<Expr>(vn)) != storage_scope_.end() &&
-        storage_scope_[GetRef<Expr>(vn)][0] != "global") {
+        storage_scope_[GetRef<Expr>(vn)].find(Expr()) != storage_scope_[GetRef<Expr>(vn)].end() &&
+        storage_scope_[GetRef<Expr>(vn)][Expr()][0] != "global") {
       Var c = Var(vn->vid, vn->type_annotation, vn->span);
       auto virtual_device = GetVirtualDevice(GetRef<Expr>(vn));
       c->virtual_device_ =
           VirtualDevice(virtual_device->device_type(), virtual_device->virtual_device_id,
-                        virtual_device->target, storage_scope_[GetRef<Expr>(vn)][0]);
+                        virtual_device->target, storage_scope_[GetRef<Expr>(vn)][Expr()][0]);
       return c;
     }
     return GetRef<Var>(vn);
   }
 
   Expr VisitExpr_(const ConstantNode* vn) final {
-    if (storage_scope_.find(GetRef<Expr>(vn)) != storage_scope_.end()) {
+    if (storage_scope_.find(GetRef<Expr>(vn)) != storage_scope_.end() &&
+        storage_scope_[GetRef<Expr>(vn)].find(Expr()) != storage_scope_[GetRef<Expr>(vn)].end()) {
       Expr c = Constant(vn->data, vn->span);
       auto virtual_device = GetVirtualDevice(GetRef<Expr>(vn));
-      c = OnDevice(c,
-                   VirtualDevice(virtual_device->device_type(), virtual_device->virtual_device_id,
-                                 virtual_device->target, storage_scope_[GetRef<Expr>(vn)][0]),
-                   true);
+      c = OnDevice(
+          c,
+          VirtualDevice(virtual_device->device_type(), virtual_device->virtual_device_id,
+                        virtual_device->target, storage_scope_[GetRef<Expr>(vn)][Expr()][0]),
+          true);
       return c;
     }
     return GetRef<Constant>(vn);
   }
 
   Expr DeviceAwareVisitExpr_(const CallNode* call_node) final {
-    auto new_call = transform::DeviceAwareExprMutator::DeviceAwareVisitExpr_(call_node);
+    // we need to duplicate ExprMutator::VisitExpr_ to correct argument scopes and
+    // put device_copy
+    auto new_op = this->Mutate(call_node->op);
+
+    tvm::Array<Type> ty_args;
+    ty_args.reserve(call_node->type_args.size());
+
+    for (auto ty_arg : call_node->type_args) {
+      auto new_ty_arg = this->VisitType(ty_arg);
+      ty_args.push_back(new_ty_arg);
+    }
+
+    tvm::Array<Expr> call_args;
+    call_args.reserve(call_node->args.size());
+    for (auto arg : call_node->args) {
+      auto new_arg = this->Mutate(arg);
+      // verification if we need to put device_copy
+      if (storage_scope_.count(arg) && storage_scope_[arg].count(GetRef<Expr>(call_node))) {
+        auto virtual_device = GetVirtualDevice(GetRef<Expr>(call_node));
+        // TOO(amalyshe): hardcoded source scope, need to fix later
```

Comment:
address this TODO - need to get memory scope here and point in the next line in generic form instead of hardcoded `global`